File: nvp/src/MVar.java
class MVar<T> {

  T content;
  boolean empty;
  Object r = new Object();
  Object w = new Object();

  public MVar() {
    content = null;
    empty = true;
  }

  public MVar(T o) {
    content = o;
    empty = false;
  }

  public T take() throws InterruptedException {
    synchronized (r) {
      while (empty) { r.wait(); }
      // here empty is false 
      synchronized (w) {
        empty = true;
        w.notify();
        return content;
      }
    }
  }

  public void put(T o) throws InterruptedException {
    synchronized (w) {
      while (!empty) { w.wait(); }
      // here empty is true
      synchronized (r) {
        empty = false;
        r.notify();
        content = o;
      }
    }
  }

  public T read()  throws InterruptedException {
    synchronized (r) {
      while (empty) { r.wait(); }
      // here empty is false 
      synchronized (w) {
        r.notify();
        return content;
      }
    }
  }
}

File: org/nvp/src/mvar.erl
-module(mvar).
-export([new_mvar/0,take_mvar/1,put_mvar/2]).

new_mvar() -> spawn(fun() -> empty() end).

empty() ->
  receive
    {put,V,P} -> P!put,
                 full(V) 
  end.

full(V) ->
  receive
    {take,P} -> P!{took,V},
                empty();
    {read,P} -> P!{read,V},
                full(V)
  end.

take_mvar(MVar) ->
  MVar!{take,self()},
  receive
     {took,V} -> V
  end.

put_mvar(MVar,V) ->
  MVar!{put,V,self()},
  receive
    put -> ok
  end.

read_mvar(Mvar) ->
  MVar!{read,self()},
  receive
    {read,V} -> V
  end.
